### ISSUE TITLE
Slow server shutdown

### ISSUE THREAD

**Issue by dodalovic** [ktorio/ktor#1819]
Hi, what would be the reason that, after I start my app's main in IntelliJ and then stop the server, it takes some seconds until it's down?

Thanks

**Comment by dmitrievanthony** [comment]
Hi @dodalovic, which server engine do you use?

**Comment by dodalovic** [comment]
Netty

**Comment by dmitrievanthony** [comment]
Hi @dodalovic, I double-checked it and it looks like it works fine. At least it matches the expectations.

The time `stop()` takes might be caused by the following. If you specify:

```
server.stop(20, 20, TimeUnit.SECONDS)
```

It means that the server is not actually forced to shut down within 20 seconds. I guess it's a lack of good documentation, but this quote from Netty doc should help:

>      * Signals this executor that the caller wants the executor to be shut down.  Once this method is called,
>      * {@link #isShuttingDown()} starts to return {@code true}, and the executor prepares to shut itself down.
>      * Unlike {@link #shutdown()}, graceful shutdown ensures that no tasks are submitted for <i>'the quiet period'</i>
>      * (usually a couple seconds) before it shuts itself down.  If a task is submitted during the quiet period,
>      * it is guaranteed to be accepted and the quiet period will start over.
>      *
>      * @param quietPeriod the quiet period as described in the documentation

Does it solve your issue?

**Comment by dodalovic** [comment]
Where should I put that? 
THis is my whole app

```kotlin
package com.odalovic

import io.ktor.application.Application
import io.ktor.application.install
import io.ktor.features.*
import io.ktor.http.HttpHeaders
import io.ktor.http.HttpMethod
import io.ktor.http.content.resources
import io.ktor.http.content.static
import io.ktor.routing.routing
import io.ktor.serialization.json

fun main(args: Array<String>): Unit = io.ktor.server.netty.EngineMain.main(args)

@Suppress("unused") // Referenced in application.conf
@kotlin.jvm.JvmOverloads
fun Application.module(testing: Boolean = false) {
    install(Compression) {
        gzip {
            priority = 1.0
        }
        deflate {
            priority = 10.0
            minimumSize(1024) // condition
        }
    }

    install(CORS) {
        method(HttpMethod.Options)
        method(HttpMethod.Put)
        method(HttpMethod.Delete)
        method(HttpMethod.Patch)
        header(HttpHeaders.Authorization)
        header("MyCustomHeader")
        allowCredentials = true
        anyHost() // @TODO: Don't do this in production if possible. Try to limit it.
    }

    install(ContentNegotiation) {
        json()
    }

    routing {
        static("/static") {
            resources("static")
        }
    }
}
```

**Comment by dmitrievanthony** [comment]
Well, if you start the server using `EngineMain` object, it intentionally assumes a default 3-sec timeout. It's unfortunately hardcoded, so as a workaround I can only suggest reimplementing the logic from this `EngineMain` in some your custom object like:

```
object CustomEngineMain {
    /**
     * Main function for starting EngineMain with Netty
     * Creates an embedded Netty application with an environment built from command line arguments.
     */
    @JvmStatic
    fun main(args: Array<String>) {
        val applicationEnvironment = commandLineEnvironment(args)
        val engine = NettyApplicationEngine(applicationEnvironment) { loadConfiguration(applicationEnvironment.config) }
        engine.addShutdownHook {
            engine.stop(0, 5, TimeUnit.SECONDS)
        }
        engine.start(true)
    }

    private fun NettyApplicationEngine.Configuration.loadConfiguration(config: ApplicationConfig) {
        val deploymentConfig = config.config("ktor.deployment")
        loadCommonConfiguration(deploymentConfig)
        deploymentConfig.propertyOrNull("requestQueueLimit")?.getString()?.toInt()?.let {
            requestQueueLimit = it
        }
        deploymentConfig.propertyOrNull("shareWorkGroup")?.getString()?.toBoolean()?.let {
            shareWorkGroup = it
        }
        deploymentConfig.propertyOrNull("responseWriteTimeoutSeconds")?.getString()?.toInt()?.let {
            responseWriteTimeoutSeconds = it
        }
    }
}
```

And then call:

```
fun main(args: Array<String>): Unit = CustomEngineMain.main(args)
```

**Comment by dmitrievanthony** [comment]
I'm closing the issue for now, but if you have any additional questions feel free to reopen it.

**Comment by ilya40umov** [comment]
It's unfortunate that every project that needs to tune this timeout needs to copy-and-paste a chunk of the framework to their codebase. Can we instead consider having a configuration property like `ktor.deployment.shutdownTimeoutSeconds`?

**Comment by FunkyMuse** [comment]
Is there any update on this matter?

As @ilya40umov said, it would be great to have that kind of configurable way.

**Comment by dodalovic** [comment]
Yes, still experiencing it

**Comment by e5l** [comment]
Fixed in `2.0.0`
